### PR TITLE
Static is rdd

### PR DESCRIPTION
**Pull request by ghislainfourny**
Depends on #452

### CODE REVIEW

File: src/main/java/sparksoniq/jsoniq/runtime/iterator/functions/base/BuiltinFunction.java [hunk 1]
```diff
@@ -6,16 +6,16 @@
 
     private FunctionIdentifier identifier;
     private final FunctionSignature signature;
-    private final Class<? extends RuntimeIterator> functionIterator;
+    private final Class<? extends RuntimeIterator> functionIteratorClass;
```
Comment (CanBerker) [inline src/main/java/sparksoniq/jsoniq/runtime/iterator/functions/base/BuiltinFunction.java]: Simple rename for clarity

File: src/main/java/sparksoniq/jsoniq/runtime/iterator/primary/ArrayRuntimeIterator.java [hunk 2]
```diff
@@ -46,7 +45,7 @@ public ArrayRuntimeIterator(RuntimeIterator arrayItems, IteratorMetadata iterato
     public Item next() {
         if (this._hasNext) {
             List<Item> result = this.runChildrenIterators(this._currentDynamicContext);
-            this._item = ItemFactory.getInstance().createArrayItem(result);
+            Item _item = ItemFactory.getInstance().createArrayItem(result);
```
Comment (CanBerker) [inline src/main/java/sparksoniq/jsoniq/runtime/iterator/primary/ArrayRuntimeIterator.java]: Variable made local

File: src/main/java/sparksoniq/jsoniq/runtime/iterator/RuntimeIterator.java [hunk 3]
```diff
@@ -191,13 +198,37 @@ public IteratorMetadata getMetadata() {
         return metadata;
     }
 
-    public abstract boolean isRDD();
+    public boolean isRDD() {
+        if (!isRDDIsSet) {
+            throw new RuntimeException("isRDD field accessed in iterator without being set.");
+        }
+        return isRDD;
```
Comment (CanBerker) [inline src/main/java/sparksoniq/jsoniq/runtime/iterator/RuntimeIterator.java]: These exceptions are put in place as safeguards to prevent unnoticed bugs. They can be potentially be removed after the implementation is checked to be sound.

File: src/main/java/sparksoniq/jsoniq/compiler/translator/expr/primary/FunctionDeclaration.java [hunk 4]
```diff
@@ -74,6 +76,20 @@ public Expression get_body() {
...
+        this.isDataFrame = false;
+
+        FunctionIdentifier identifier = new FunctionIdentifier(this._name, this._params.size());
+        // if named(static) function declaration
+        if (!this._name.equals("")) {
+            // TODO: find a way to set isRDD and isDF from this._body
+            Functions.addUserDefinedFunctionIsRDD(identifier, false, this.getMetadata());
+            Functions.addUserDefinedFunctionIsDataFrame(identifier, false, this.getMetadata());
+        }
+    }
```
Comment (CanBerker) [inline src/main/java/sparksoniq/jsoniq/compiler/translator/expr/primary/FunctionDeclaration.java]: Note: This is still not addressed

File: src/main/java/sparksoniq/jsoniq/runtime/iterator/DataFrameRuntimeIterator.java [hunk 5]
```diff
@@ -0,0 +1,34 @@
...
+ *
+ */
+
+package sparksoniq.jsoniq.runtime.iterator;
+
+import sparksoniq.jsoniq.runtime.metadata.IteratorMetadata;
+
+import java.util.List;
+
+public abstract class DataFrameRuntimeIterator extends RDDRuntimeIterator {
```
Comment (CanBerker) [inline src/main/java/sparksoniq/jsoniq/runtime/iterator/DataFrameRuntimeIterator.java]: Class defined for DataFrame only Iterators. eg. Parquet file and Structured Json File.

In the old implementation this would have a method isDataFrame which would always return true.

New implementation uses this class while checking if a function's iterator is a subclass of this.

File: src/main/resources/test_files/runtime-spark/Pushdown/KeysDF1.jq [hunk 6]
```diff
@@ -1,2 +1,2 @@
-(:JIQS: ShouldRun; Output="(choices, country, date, guess, sample, target)" :)
+(:JIQS: ShouldRun; Output="(country, choices, guess, target, date, sample)" :)
 keys(structured-json-file("./src/main/resources/queries/conf-ex.json"))
```
Comment (CanBerker) [inline src/main/resources/test_files/runtime-spark/Pushdown/KeysDF1.jq]: This test is executed locally on the master which I believe is not correct. It is now executed with RDD API on the keys iterator which gives the new results.

File: src/main/java/sparksoniq/jsoniq/compiler/translator/expr/flowr/ForClause.java [hunk 7]
```diff
@@ -38,12 +38,28 @@ public ForClause(List<ForClauseVar> vars, ExpressionMetadata metadataFromContext
...
         return forVariables;
     }
 
+    @Override
+    public void setPreviousClause(FlworClause previousClause) {
+        super.setPreviousClause(previousClause);
+    }
+
+    @Override
+    protected void initIsRDD() {
```
[the remark this answers is not in the document]
Comment (CanBerker) [reply src/main/java/sparksoniq/jsoniq/compiler/translator/expr/flowr/ForClause.java]: Done

File: src/main/java/sparksoniq/jsoniq/compiler/translator/expr/flowr/ForClause.java [hunk 8]
```diff
@@ -38,12 +38,28 @@ public ForClause(List<ForClauseVar> vars, ExpressionMetadata metadataFromContext
...
+        for (int varIndex = forVariables.size() - 1; varIndex > 0; varIndex--) {
+            forVariables.get(varIndex).setPreviousClause(forVariables.get(varIndex - 1));
+        }
     }
 
     public List<ForClauseVar> getForVariables() {
         return forVariables;
     }
 
+    @Override
```
[the remark this answers is not in the document]
Comment (CanBerker) [reply src/main/java/sparksoniq/jsoniq/compiler/translator/expr/flowr/ForClause.java]: I believe this was an oversight. I have updated the behavior similar to the letClause as both of these clauses can have a multitude of variable definitions, which must be chained together for correctness.

File: src/main/java/sparksoniq/jsoniq/compiler/translator/expr/primary/VariableReference.java [hunk 9]
```diff
@@ -38,6 +38,12 @@
     public VariableReference(String _name, ExpressionMetadata metadata) {
         super(metadata);
         this.name = _name;
+        // TODO: can we eliminate this assignment by properly visiting VariableReferences in staticContextVisitor
+        _highestExecutionMode = ExecutionMode.LOCAL;
```
Comment (CanBerker) [inline src/main/java/sparksoniq/jsoniq/compiler/translator/expr/primary/VariableReference.java]: Removing this line causes a multitude of runtime exceptions as VariableReferenceIterators(created from variableReferences that inherit from Expression) have their execution mode unset. This is caused by them not being properly visited by the staticContextVisitor which sets the execution mode by looking up the variable in the context. I think this will require a bit of work on the staticContextVisitor if we wish to address this.

File: src/main/java/sparksoniq/semantics/visitor/StaticContextVisitor.java [hunk 10]
```diff
@@ -48,76 +49,109 @@
...
+        // define a static context for the function body, add params to the context and visit the body expression
+        StaticContext functionDeclarationContext = new StaticContext(argument);
+        expression.get_params()
+            .forEach(
+                (paramName, flworVarSequenceType) -> functionDeclarationContext.addVariable(
+                    paramName,
+                    flworVarSequenceType.getSequence(),
+                    expression.getMetadata(),
+                    ExecutionMode.LOCAL // static udf currently supports materialized(local) params, not RDDs or DFs
+                )
```
Comment (CanBerker) [inline src/main/java/sparksoniq/semantics/visitor/StaticContextVisitor.java]: At the moment, static function declarations allow parametrized variables to be in the materialized form. In other words, (and if I'm not mistaken), user defined functions would not work properly when a variable that is a DF or an RDD is passed as a parameter.

File: src/main/java/sparksoniq/spark/iterator/flowr/ForClauseSparkIterator.java [hunk 11]
```diff
@@ -54,30 +54,25 @@
 
     private static final long serialVersionUID = 1L;
     private String _variableName; // for efficient use in local iteration
-    private RuntimeIterator _expression;
+    private RuntimeIterator _assignmentIterator;
```
Comment (CanBerker) [inline src/main/java/sparksoniq/spark/iterator/flowr/ForClauseSparkIterator.java]: Simple renaming for the iterators used in for and let clause iterators

File: src/main/java/sparksoniq/spark/iterator/flowr/ForClauseSparkIterator.java [hunk 12]
```diff
@@ -54,30 +54,25 @@
...
     private Map<String, DynamicContext.VariableDependency> _dependencies;
     private DynamicContext _tupleContext; // re-use same DynamicContext object for efficiency
     private FlworTuple _nextLocalTupleResult;
     private FlworTuple _inputTuple; // tuple received from child, used for tuple creation
 
     public ForClauseSparkIterator(
             RuntimeTupleIterator child,
-            VariableReferenceIterator variableReference,
-            RuntimeIterator assignmentExpression,
+            String variableName,
```
Comment (CanBerker) [inline src/main/java/sparksoniq/spark/iterator/flowr/ForClauseSparkIterator.java]: Building an iterator just to get the name of the variable was redundant for 'for' and 'let' clauses. We simply pass in the variable name now

File: src/main/java/sparksoniq/jsoniq/runtime/iterator/primary/VariableReferenceIterator.java [hunk 13]
```diff
@@ -44,19 +45,12 @@
     private List<Item> _items = null;
     private int _currentIndex = 0;
 
-    public VariableReferenceIterator(String variableName, SequenceType seq, IteratorMetadata iteratorMetadata) {
-        super(null, iteratorMetadata);
-        _variableName = "$" + variableName;
+    public VariableReferenceIterator(String variableName, SequenceType seq, ExecutionMode executionMode, IteratorMetadata iteratorMetadata) {
+        super(null, executionMode, iteratorMetadata);
+        _variableName = variableName;
```
Comment (CanBerker) [inline src/main/java/sparksoniq/jsoniq/runtime/iterator/primary/VariableReferenceIterator.java]: In order to be consistent across the application, I have standardized the variable names in wherever they are used. Variables are now stored and referenced in the dynamic context without the $ prefix. They are also mentioned in error messages without the prefix. I have opted for this as the behavior now matches Zorba.

File: src/main/java/sparksoniq/jsoniq/runtime/iterator/functions/base/Functions.java [hunk 14]
```diff
@@ -522,44 +524,66 @@ public static boolean checkBuiltInFunctionExists(FunctionIdentifier identifier)
...
+                                IteratorMetadata.class
+                        );
+                functionCallIterator = constructor.newInstance(arguments, executionMode, metadata);
+            } catch (ReflectiveOperationException ex) {
+                throw new UnknownFunctionCallException(
+                        identifier.getName(),
+                        arguments.size(),
+                        metadata.getExpressionMetadata()
+                );
+            }
```
Comment (CanBerker) [inline src/main/java/sparksoniq/jsoniq/runtime/iterator/functions/base/Functions.java]: I'd like to have your feedback on this one @ghislainfourny. To minimize the refactoring, I have initially opted for adding the execution mode as a parameter to the HybridRuntimeIterator and its descendants. LocalRuntimeIterators simply get this parameter fixed to LOCAL. This resulted in hybrid enabled built-in functions having constructors with 3 parameters while Local built-in functions have constructors with 2 parameters. In order to handle this and get the tests running. I have implemented the above fix which is quick and dirty.

One potential solution that would still be somewhat dirty but more performant. We could eliminate the exceptions by performing a certain check on the constructor to avoid performance loss due to exception handling.

Alternative feels more neat however will require a bit more refactoring. We can simply add ExecutionMode to all the constructors of all iterators and enforce that localRuntimeIterators can only be assigned ExecutionMode.LOCAL. A similar implementation can be found in DataFrameRuntimeIterator or RDDRuntimeIterator. 

What is your take on this? Do you see any preferable alternatives/suggestions?

File: src/main/java/sparksoniq/jsoniq/compiler/translator/expr/primary/FunctionDeclaration.java (hunk 4 above)
Comment (CanBerker) [reply src/main/java/sparksoniq/jsoniq/compiler/translator/expr/primary/FunctionDeclaration.java]: Outdated

File: src/main/java/sparksoniq/spark/iterator/flowr/ForClauseSparkIterator.java (hunk 12 above)
Comment (CanBerker) [reply src/main/java/sparksoniq/spark/iterator/flowr/ForClauseSparkIterator.java]: Outdated

File: src/main/java/sparksoniq/jsoniq/compiler/translator/expr/flowr/FlworClause.java [hunk 15]
```diff
@@ -50,8 +54,9 @@ public void setPreviousClause(FlworClause previousClause) {
         this.previousClause = previousClause;
     }
 
-    public FLWOR_CLAUSES getClauseType() {
-        return clauseType;
+    @Override
+    public void initHighestExecutionMode() {
+        this._highestExecutionMode = previousClause.getHighestExecutionMode();
```
Comment (ghislainfourny) [inline src/main/java/sparksoniq/jsoniq/compiler/translator/expr/flowr/FlworClause.java]: Could we add a comment that this is meant to be overridden in some cases? Thanks!

File: src/main/java/sparksoniq/jsoniq/compiler/translator/expr/flowr/FlworExpression.java [hunk 16]
```diff
@@ -77,6 +77,12 @@ private void set_returnClause(ReturnClause returnClause) {
     }
 
     @Override
+    public ExecutionMode getHighestExecutionMode() {
+        this._highestExecutionMode = ExecutionMode.UNSET;
```
Comment (ghislainfourny) [inline src/main/java/sparksoniq/jsoniq/compiler/translator/expr/flowr/FlworExpression.java]: A side effect in a get function raises red flags. Is the default not unset anyway (because of init), and do we not just ignore it?

File: src/main/java/sparksoniq/jsoniq/compiler/translator/expr/flowr/ForClause.java [hunk 17]
```diff
@@ -56,6 +57,15 @@ public void setPreviousClause(FlworClause previousClause) {
         forVariables.get(0).previousClause = this.previousClause;
     }
 
+    @Override
+    public void initHighestExecutionMode() {
+        // call isDataFrame on the last forVariable
+        this._highestExecutionMode =
+            forVariables.get(forVariables.size() - 1).getHighestExecutionMode().isDataFrame()
+                ? ExecutionMode.DATAFRAME
+                : ExecutionMode.LOCAL;
```
Comment (ghislainfourny) [inline src/main/java/sparksoniq/jsoniq/compiler/translator/expr/flowr/ForClause.java]: Is there a reason why not just

this._highestExecutionMode = forVariables.get(forVariables.size() - 1).getHighestExecutionMode()

?

File: src/main/java/sparksoniq/jsoniq/compiler/translator/expr/flowr/LetClause.java [hunk 18]
```diff
@@ -55,6 +56,15 @@ public void setPreviousClause(FlworClause previousClause) {
         letVars.get(0).previousClause = this.previousClause;
     }
 
+    @Override
+    public void initHighestExecutionMode() {
+        // call isDataFrame on the last letVariable
+        this._highestExecutionMode =
+            letVars.get(letVars.size() - 1).getHighestExecutionMode().isDataFrame()
+                ? ExecutionMode.DATAFRAME
```
Comment (ghislainfourny) [inline src/main/java/sparksoniq/jsoniq/compiler/translator/expr/flowr/LetClause.java]: The same applies here: why not

this._highestExecutionMode = letVars.get(letVars.size() - 1).getHighestExecutionMode()

?

File: src/main/java/sparksoniq/jsoniq/compiler/translator/expr/flowr/OrderByClauseExpr.java [hunk 19]
```diff
@@ -48,6 +49,12 @@ public OrderByClauseExpr(
         this._emptyOrder = empty_order;
     }
 
+    @Override
+    public void initHighestExecutionMode() {
+        // OrderByClauseExpr's execution mode is not used. Leave it unset
+        this._highestExecutionMode = ExecutionMode.UNSET;
```
Comment (ghislainfourny) [inline src/main/java/sparksoniq/jsoniq/compiler/translator/expr/flowr/OrderByClauseExpr.java]: Why is it not used, i.e., why is this clause treated differently?

File: src/main/java/sparksoniq/jsoniq/tuple/FlworTuple.java [hunk 20]
```diff
@@ -104,7 +102,7 @@ public boolean isRDD(String key, IteratorMetadata metadata) {
             || dfVariables.containsKey(key);
     }
 
-    public boolean isDF(String key, IteratorMetadata metadata) {
+    public boolean isDataFrame(String key, IteratorMetadata metadata) {
```
Comment (ghislainfourny) [inline src/main/java/sparksoniq/jsoniq/tuple/FlworTuple.java]: This can be moved to a separate PR, reviewed and merged.

File: src/main/java/sparksoniq/jsoniq/compiler/translator/expr/postfix/extensions/DynamicFunctionCallExtension.java [hunk 21]
```diff
@@ -48,6 +49,11 @@ public DynamicFunctionCallExtension(List<Expression> arguments, ExpressionMetada
         return getDescendantsFromChildren(result, depthSearch);
     }
 
+    @Override
+    public void initHighestExecutionMode() {
+        this._highestExecutionMode = ExecutionMode.LOCAL;
```
Comment (ghislainfourny) [inline src/main/java/sparksoniq/jsoniq/compiler/translator/expr/postfix/extensions/DynamicFunctionCallExtension.java]: Is it not the default already, i.e., is this not superfluous?

File: src/main/java/sparksoniq/jsoniq/compiler/translator/expr/flowr/FlworExpression.java (hunk 16 above)
Comment (CanBerker) [reply src/main/java/sparksoniq/jsoniq/compiler/translator/expr/flowr/FlworExpression.java]: You're absolutely right. That was simply an oversight on my part. That line was meant for the init method.

File: src/main/java/sparksoniq/jsoniq/compiler/translator/expr/flowr/ForClause.java (hunk 17 above)
Comment (CanBerker) [reply src/main/java/sparksoniq/jsoniq/compiler/translator/expr/flowr/ForClause.java]: No specific reason, just an oversight :) Good catch!

File: src/main/java/sparksoniq/jsoniq/compiler/translator/expr/flowr/LetClause.java (hunk 18 above)
Comment (CanBerker) [reply src/main/java/sparksoniq/jsoniq/compiler/translator/expr/flowr/LetClause.java]: Also fixed here

File: src/main/java/sparksoniq/jsoniq/compiler/translator/expr/postfix/extensions/DynamicFunctionCallExtension.java (hunk 21 above)
Comment (CanBerker) [reply src/main/java/sparksoniq/jsoniq/compiler/translator/expr/postfix/extensions/DynamicFunctionCallExtension.java]: Postfix extension (parent class of DynamicFunctionCallExtension) overrides init method to the behavior which is shared by all postfix extensions except dynamicFunctionCallExtension. This is always local, which is the behavior identical to the grandparent.

File: src/main/java/sparksoniq/jsoniq/tuple/FlworTuple.java (hunk 20 above)
Comment (CanBerker) [reply src/main/java/sparksoniq/jsoniq/tuple/FlworTuple.java]: Done. Please review at #446

File: src/main/java/sparksoniq/jsoniq/compiler/translator/expr/flowr/FlworClause.java (hunk 15 above)
Comment (CanBerker) [reply src/main/java/sparksoniq/jsoniq/compiler/translator/expr/flowr/FlworClause.java]: Added

File: src/main/java/sparksoniq/jsoniq/compiler/translator/expr/postfix/extensions/DynamicFunctionCallExtension.java (hunk 21 above)
Comment (CanBerker) [reply src/main/java/sparksoniq/jsoniq/compiler/translator/expr/postfix/extensions/DynamicFunctionCallExtension.java]: I have added a comment to the code as well

File: src/main/java/sparksoniq/jsoniq/compiler/translator/expr/flowr/OrderByClauseExpr.java (hunk 19 above)
Comment (CanBerker) [reply src/main/java/sparksoniq/jsoniq/compiler/translator/expr/flowr/OrderByClauseExpr.java]: This again due to the current nature of how we are creating and using our iterators. These orderbyClauseExpressions are the individual expressions within an order by expression (ie. order by $x asc, $y dec.... ). The execution mode of such individual expressions are not used in the current state of the program. 

Overall orderByClause gets its execution mode set based on the previousClause and this expression's execution mode sets the execution mode of the OrderByClauseSparkIterator. 

The orderByClauseExpr (the subexpressions) contain the actual expression along with the ascending/descending and empty handling information. This actual expression is what the iterator is generated from and it sets the execution mode directly for the iterator. As orderByClauseExpr does not partake in generation and execution mode of an iterator, its execution mode is unused in the current implementation.

I generally opt for setting these UNSET values explicitly in the initialization for showing the deliberate nature of this value, especially for the case if the implementation changes in the future and orderByClauseExpr plays a more direct role for a RuntimeIterator.

File: src/main/java/sparksoniq/jsoniq/runtime/iterator/primary/VariableReferenceIterator.java (hunk 13 above)
Comment (CanBerker) [reply src/main/java/sparksoniq/jsoniq/runtime/iterator/primary/VariableReferenceIterator.java]: Outdated

File: src/main/java/sparksoniq/jsoniq/runtime/iterator/HybridRuntimeIterator.java [hunk 22]
```diff
@@ -40,26 +40,25 @@
...
+        return false;
     }
 
-    @Override
-    public boolean isDataFrame() {
-        return false;
+    protected void fallbackToRDDIfDFNotImplemented(ExecutionMode executionMode) {
+        if (executionMode == ExecutionMode.DATAFRAME && !this.implementsDataFrames()) {
+            this._highestExecutionMode = ExecutionMode.RDD;
+        }
```
Comment (CanBerker) [inline src/main/java/sparksoniq/jsoniq/runtime/iterator/HybridRuntimeIterator.java]: I will investigate again why this function was needed in the first place, it may be masking some other correctness issue

File: src/main/java/sparksoniq/jsoniq/compiler/translator/expr/flowr/FlworClause.java (hunk 15 above)
Comment (ghislainfourny) [reply src/main/java/sparksoniq/jsoniq/compiler/translator/expr/flowr/FlworClause.java]: Thanks!

File: src/main/java/sparksoniq/jsoniq/compiler/translator/expr/flowr/FlworExpression.java (hunk 16 above)
Comment (ghislainfourny) [reply src/main/java/sparksoniq/jsoniq/compiler/translator/expr/flowr/FlworExpression.java]: Thanks!

File: src/main/java/sparksoniq/jsoniq/compiler/translator/expr/primary/FunctionCall.java [hunk 23]
```diff
@@ -52,11 +62,64 @@ public String getFunctionName() {
...
     }
 
+    @Override
+    public final void initHighestExecutionMode() {
+        throw new OurBadException("Function call expressions do not use the highestExecutionMode initializer");
+    }
+
+    public void initFunctionCallHighestExecutionMode(boolean ignoreMissingFunctionError) {
+        FunctionIdentifier identifier = new FunctionIdentifier(this._functionName, this._arguments.size());
+        if (Functions.checkBuiltInFunctionExists(identifier)) {
```
Comment (ghislainfourny) [inline src/main/java/sparksoniq/jsoniq/compiler/translator/expr/primary/FunctionCall.java]: My intuition tells me that this code could be simplified by adding a return statement after each _highestExecutionMode assignments. Then all elses would go away and the structure would become flatter and easier to read.

File: src/main/java/sparksoniq/semantics/visitor/VisitorHelpers.java [hunk 24]
```diff
@@ -0,0 +1,19 @@
...
+import sparksoniq.jsoniq.runtime.iterator.RuntimeIterator;
+import sparksoniq.semantics.StaticContext;
+
+public class VisitorHelpers {
+
+    public static RuntimeIterator generateRuntimeIterator(ExpressionOrClause expression, RuntimeIterator argument) {
+        return new RuntimeIteratorVisitor().visit(expression, argument);
+    }
+
+    public static StaticContext generateStaticContextDoublePass(ExpressionOrClause expression, StaticContext argument) {
```
Comment (ghislainfourny) [inline src/main/java/sparksoniq/semantics/visitor/VisitorHelpers.java]: public static void populateStaticContext(ExpressionOrClause expression)

File: src/main/java/sparksoniq/semantics/visitor/VisitorHelpers.java [hunk 25]
```diff
@@ -0,0 +1,19 @@
+package sparksoniq.semantics.visitor;
+
+import sparksoniq.jsoniq.compiler.translator.expr.ExpressionOrClause;
+import sparksoniq.jsoniq.runtime.iterator.RuntimeIterator;
+import sparksoniq.semantics.StaticContext;
+
+public class VisitorHelpers {
+
+    public static RuntimeIterator generateRuntimeIterator(ExpressionOrClause expression, RuntimeIterator argument) {
```
Comment (ghislainfourny) [inline src/main/java/sparksoniq/semantics/visitor/VisitorHelpers.java]: public static RuntimeIterator generateRuntimeIterator(ExpressionOrClause expression)

 (argument is null always)

File: src/main/java/sparksoniq/semantics/visitor/VisitorHelpers.java [hunk 26]
```diff
@@ -0,0 +1,19 @@
...
+        return new RuntimeIteratorVisitor().visit(expression, argument);
+    }
+
+    public static StaticContext generateStaticContextDoublePass(ExpressionOrClause expression, StaticContext argument) {
+        StaticContextVisitor visitor = new StaticContextVisitor();
+        visitor.visit(expression, argument);
```
Comment (ghislainfourny) [inline src/main/java/sparksoniq/semantics/visitor/VisitorHelpers.java]: argument is null always

File: src/main/java/sparksoniq/semantics/visitor/VisitorHelpers.java [hunk 27]
```diff
@@ -0,0 +1,19 @@
...
+        visitor.setConfigForConsequentPasses();
+        return visitor.visit(expression, argument);
```
Comment (ghislainfourny) [inline src/main/java/sparksoniq/semantics/visitor/VisitorHelpers.java]: argument is null always

File: src/main/java/sparksoniq/jsoniq/compiler/translator/expr/primary/FunctionCall.java (hunk 23 above)
Comment (ghislainfourny) [reply src/main/java/sparksoniq/jsoniq/compiler/translator/expr/primary/FunctionCall.java]: Offline discussion: function catalogue maps each builtin function to execution mode, and here we only look it up.

File: src/main/java/sparksoniq/semantics/visitor/VisitorHelpers.java (hunk 25 above)
Comment (CanBerker) [reply src/main/java/sparksoniq/semantics/visitor/VisitorHelpers.java]: Done

File: src/main/java/sparksoniq/semantics/visitor/VisitorHelpers.java (hunk 24 above)
Comment (CanBerker) [reply src/main/java/sparksoniq/semantics/visitor/VisitorHelpers.java]: Done

File: src/main/java/sparksoniq/semantics/visitor/VisitorHelpers.java (hunk 26 above)
Comment (CanBerker) [reply src/main/java/sparksoniq/semantics/visitor/VisitorHelpers.java]: Done

File: src/main/java/sparksoniq/semantics/visitor/VisitorHelpers.java (hunk 27 above)
Comment (CanBerker) [reply src/main/java/sparksoniq/semantics/visitor/VisitorHelpers.java]: Done

File: src/main/java/sparksoniq/jsoniq/runtime/iterator/functions/base/Functions.java (hunk 14 above)
Comment (CanBerker) [reply src/main/java/sparksoniq/jsoniq/runtime/iterator/functions/base/Functions.java]: TODO - Offline discussion: expose executionMode parameters from all iterator constructors, remove the exception handling Above

File: src/main/java/sparksoniq/jsoniq/compiler/translator/expr/primary/FunctionCall.java (hunk 23 above)
Comment (CanBerker) [reply src/main/java/sparksoniq/jsoniq/compiler/translator/expr/primary/FunctionCall.java]: I agree with the simplification using returns and I will implement this.

Regarding the second point: for the case of LocalRuntimeIterator, RDDRuntimeIterators or DataFrameRuntimeIterators where execution mode is always a fixed value, the catalogue approach would work. However in the case of HybridRuntimeIterators we need some custom logic in place that determines the execution mode based on the the child expression. But even this logic is not shared by all hybrid iterators due to peculiarities of individual iterators such as object 'keys', where we resort to local execution also when child is a DataFrame. 

Given this information that not all built-in functions have a static executionMode that can be accessed with a simple lookup, how would you like to proceed?

Comment (CanBerker) [reply src/main/java/sparksoniq/jsoniq/compiler/translator/expr/primary/FunctionCall.java]: The return oriented implementation is in place. I have also attempted to tidy up the code a bit to increase readability. 

I am looking forward to your feedback for further improvements or alternative solutions to the mentioned issue.

Comment (ghislainfourny) [reply src/main/java/sparksoniq/jsoniq/compiler/translator/expr/primary/FunctionCall.java]: Many thanks for the clean return-based implementation.

I am optimistic on the hybrid function iterators, especially looking at the cleaned code below, because there are not so many cases and we can come up with an larger list of function-specific execution modes (local, RDD, DF, but also same-as-first-parameter, etc). Then we just need to map this larger list (stored and looked up in the catalogue) to our "regular" execution mode list looking at the parameter iterators.

Could you please prepare a list of all functions that are based on a hybrid iterator? Then we'll take a closer look at them.

Thanks!

File: src/main/java/sparksoniq/jsoniq/runtime/iterator/functions/DynamicFunctionCallIterator.java [hunk 28]
```diff
@@ -123,6 +125,7 @@ private void setFunctionItemAndIteratorWithCurrentContext() {
         }
         _functionCallIterator = Functions.buildUserDefinedFunctionCallIterator(
             _functionItem,
+            _functionItem.getBodyIterator().getHighestExecutionMode(),
```
Comment (ghislainfourny) [inline src/main/java/sparksoniq/jsoniq/runtime/iterator/functions/DynamicFunctionCallIterator.java]: For dynamic call, I think this is always local no matter what.

(only static UDF calls are able to execute the function body as an RDD).

**Comment by ghislainfourny** [comment]
@CanBerker this is going in the right direction. I only have the two visitors as well as the function call mechanism left to review and we can fine-tune the mechanism there (see comment).

File: src/main/java/sparksoniq/jsoniq/runtime/iterator/functions/base/Functions.java (hunk 14 above)
Comment (CanBerker) [reply src/main/java/sparksoniq/jsoniq/runtime/iterator/functions/base/Functions.java]: Done

File: src/main/java/sparksoniq/spark/iterator/flowr/ForClauseSparkIterator.java (hunk 11 above)
Comment (CanBerker) [reply src/main/java/sparksoniq/spark/iterator/flowr/ForClauseSparkIterator.java]: Outdated

File: src/main/java/sparksoniq/jsoniq/compiler/translator/expr/primary/VariableReference.java (hunk 9 above)
Comment (CanBerker) [reply src/main/java/sparksoniq/jsoniq/compiler/translator/expr/primary/VariableReference.java]: This is addressed

File: src/main/java/sparksoniq/jsoniq/compiler/translator/expr/primary/FunctionCall.java (hunk 23 above)
Comment (CanBerker) [reply src/main/java/sparksoniq/jsoniq/compiler/translator/expr/primary/FunctionCall.java]: These are the Hybrid function iterators:
- keys
- values
- distinct-values

- json-file
- parallelize
- text-file

Do you wish to perform this refactoring within this PR or shall we address introduction of function executionMode catalogue and functionIterator-only execution modes in a separate PR?

File: src/main/java/sparksoniq/jsoniq/runtime/iterator/functions/DynamicFunctionCallIterator.java (hunk 28 above)
Comment (CanBerker) [reply src/main/java/sparksoniq/jsoniq/runtime/iterator/functions/DynamicFunctionCallIterator.java]: DynamicFunctionCallRuntimeIterator always evaluates locally indeed. However if a UDF is RDD/DF enabled, I think it could be evaluated in that manner. In this scope of this PR all UDFs are locally evaluated indeed. But #451 will be bringing the support for RDD/DF enabled UDFs.
For the example below,  statically detected postFixExtension of dynamicFunctionCall will be locally executed. However the function invoked by this dynamicFunctionCall could be evaluated with RDDs. Please correct me if there is something that I am missing.

declare function test() as integer{
    parallelize((1,2,3))
};
test#0()

File: src/main/java/sparksoniq/jsoniq/compiler/translator/expr/primary/FunctionCall.java (hunk 23 above)
Comment (ghislainfourny) [reply src/main/java/sparksoniq/jsoniq/compiler/translator/expr/primary/FunctionCall.java]: Is it not just keys, values and distinct-values?

Among these: json-file, parallelize and text-file are anyway always RDDs (RDDRuntimeIterator), correct?

So if it is just keys, values and distinct-values that are "truly" hybrid and could be either local or RDD, then it should be quite easy as we only need one more case: when the highestExecutionMode is taken from the first parameter's highestExecutionMode. All three functions work that way.

This can be done with a separate, function-specific enum.

Comment (CanBerker) [reply src/main/java/sparksoniq/jsoniq/compiler/translator/expr/primary/FunctionCall.java]: Yes absolutely, that was a mistake. It's only the three that you have mentioned.

The cases are accurate. I will implement these next.

File: src/main/java/sparksoniq/semantics/visitor/StaticContextVisitor.java (hunk 10 above)
Comment (ghislainfourny) [reply src/main/java/sparksoniq/semantics/visitor/StaticContextVisitor.java]: I agree with always passing materialized parameters to UDFs. This makes the higherExecutionMode of the expressions in the function body not depend on the parameters, and "truly static": indeed, if the same function is called at multiple places, there is no guarantee the passed parameters will always be local or always be RDDs.

If we relax this later, this will require significantly more work to get right; now would be too premature.

**Comment by ghislainfourny** [comment]
@CanBerker I have now looked at the visitors and it looks fine too.

Thus, to the best of my knowledge, the builtin function "extended execution mode map" (if we call it that way) is the last issue to handle :)

File: src/main/java/sparksoniq/jsoniq/runtime/iterator/HybridRuntimeIterator.java (hunk 22 above)
Comment (ghislainfourny) [reply src/main/java/sparksoniq/jsoniq/runtime/iterator/HybridRuntimeIterator.java]: I think that's fine for now -- in the future, it would be desirable to have a perfect match between what the static analysis says and what actually happens at runtime, but we can fine-tune later.

**Comment by ghislainfourny** [comment]
Thanks @CanBerker also for the URL validation. That's good to go! (I can't approve the PR as I created it).
